Update the eip to store in ring buffer instead of serial storage and add eip 158 handling strategy (#3)
gballet/EIPs: EIPS/eip-2935.md

## Motivation

Currently BLOCKHASH opcode accesses history to resolve hash of the block number in EVM. However a more stateless client friendly way is to maintain and serve these hashes from state.

Although this is possible even in Merkle trie state, but Verkle trie state further allows bundling the BLOCKHASH witnesses (along with other witnesses) in an efficient manner making it worthwhile to have these in state.

A side benefit of this approach could be that it allows building/validating proofs related to last 256 ancestors directly against the current state.

## Specification

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP` | TBD |
| `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
| `HISTORY_SERVE_WINDOW` | `256` |

This EIP specifies for storing last `HISTORY_SERVE_WINDOW` block hashes in a ring buffer storage of `HISTORY_SERVE_WINDOW` length.

At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:

```python
def process_block_hash_history(block :Block, state: State):

>     if block.timestamp >= FORK_TIMESTAMP:

Reviewer: ```suggestion
    if block.timestamp >= FORK_TIMESTAMP and block.number > 0:
```

Author: ummm  isn't this a bit redundant? genesis block anyway doesn't undergo processing

>         state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1 % HISTORY_SERVE_WINDOW , block.parent.hash)

Reviewer: ```suggestion
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)
```

## Motivation

Currently BLOCKHASH opcode accesses history to resolve hash of the block number in EVM. However a more stateless client friendly way is to maintain and serve these hashes from state.

Although this is possible even in Merkle trie state, but Verkle trie state further allows bundling the BLOCKHASH witnesses (along with other witnesses) in an efficient manner making it worthwhile to have these in state.

A side benefit of this approach could be that it allows building/validating proofs related to last 256 ancestors directly against the current state.

## Specification

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP` | TBD |
| `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
| `HISTORY_SERVE_WINDOW` | `256` |

This EIP specifies for storing last `HISTORY_SERVE_WINDOW` block hashes in a ring buffer storage of `HISTORY_SERVE_WINDOW` length.

At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:

```python

> def process_block_hash_history(block :Block, state: State):

Reviewer 2: ```suggestion
def process_block_hash_history(block: Block, state: State):
```

    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(HISTORY_SERVE_WINDOW - 1):
            # stop at genesis block
            if ancestor.number == 0:
                break

            ancestor = ancestor.parent
            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number % HISTORY_SERVE_WINDOW, ancestor.hash)
```

Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` ancestors (up until genesis).

For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp >= FORK_TIMESTAMP`), switch the logic to:

```python
def resolve_blockhash(block: Block, state: State, arg: uint64):
  # check the wrap around range

>   if arg >= block.number or arg < max(block.number - HISTORY_SERVE_WINDOW, 0)

Reviewer 3: Note: EVM cannot have numbers below 0

    return 0

  return state.load_slot(HISTORY_STORAGE_ADDRESS, arg % HISTORY_SERVE_WINDOW)
```

Some activation scenarios:

 * For the fork to be activated at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
 * for activation at block `1`, only genesis hash will be written at slot `0` as there is no additional history that needs to be persisted.
 * for activation at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.

>  * for activation at block `1000`, block `744-999`'s hashes will be presisted in the slot and `BLOCKHASH` for `733` or less would resolve to `0` as only `HISTORY_SERVE_WINDOW` can be served.

Reviewer 3: ```suggestion
 * for activation at block `1000`, block `744-999`'s hashes will be presisted in the slot and `BLOCKHASH` for `743` or less would resolve to `0` as only `HISTORY_SERVE_WINDOW` can be served.
```

I think the numbers here are incorrect?

### [EIP-158](./eip-158.md) handling

This address is currently exempt from [EIP-158](./eip-158.md) cleanup in Kaustinen Verkle Testnet but we plan to address this in the following way:

* Deploy a contract à la [EIP-4788](./eip-4788.md) which just supports `get` method to resolve the BLOCKHASH as per the logic defined above (and use the generated address as the BLOCKHASH contract address).

> * While the clients are expected to directly read from state (or maintain and serve from memory) to resolve BLOCKHASH opcode, this contract's `get` could be invoked by transaction (via another contract or directly) leading to a normal contract execution (and gas consumption) as per the semantics of the contract call.

Reviewer 3: This is fine, but as note: we will need a similar process like 4788 regarding auditing this contract etc etc in order to ship this :)

Reviewer 3: With 4788 there are also checks in the contract (if timestamp does not match storage, then throw). We now seem to treat this contract on a different level: where is the "requested block hash >= block.number - 256" check? Is this in the contract, or is this done "outside" the contract? This is not clear from the current description (at least, for me)

Author: contract will implement `resolve_blockhash` logic so yes the check goes inside.

4788 worries about timestamp because could be missed slots, but no such thing as missed numbers (if i understand correctly)

Reviewer 3: Yes to be clear: if one would call `resolve_blockhash(n)` and then calldata `n` would just return `sload(n % HISTORY_SERVE_WINDOW)` then we would be in trouble, but if there is this check in the contract that we are within range (i.e. we are within 256 blocks) then we are fine. I think I would just like an explicit note here such that the "shortcut" implementation in clients (i.e. not call the contract but write/read to storage immediately) has a consistent check in the EVM contract :smile:

Author: there is a check there, could you confirm if that is correct?

Reviewer 3: If the check is also in the EVM bytecode then it looks correct!

## Rationale

Very similar ideas were proposed before in [EIP-210](./eip-210.md) et al. This EIP is a simplification, removing two sources of needless complexity:

1. Having a tree-like structure with multiple layers as opposed to a single list
2. Writing the EIP in EVM code

> 3. Serial unbounded storage of hashes for a deep access to the history

Reviewer 3: I'm not sure does this comment apply? Can only get latest 256 blockhashes per this change? I might interpret this comment wrong though?

Author: yes correct, so we considered serial unbounded storage but only went with ring buffer

Reviewer 3: Ah right sorry, I did not see that this was in the "simplification" part (thus removing the unbounded storage by this PR)

> However after weighing pros and cons, we decided to go with just a limited ring buffer to only serve the requiste `HISTORY_SERVE_WINDOW` as [EIP-4788](./eip-4788.md) and beacon state accumulators allow (albit a bit more complex) proof against any ancestor since merge.

Reviewer 3: ```suggestion
However after weighing pros and cons, we decided to go with just a limited ring buffer to only serve the requisite `HISTORY_SERVE_WINDOW` as [EIP-4788](./eip-4788.md) and beacon state accumulators allow (albeit a bit more complex) proof against any ancestor since merge.
```

Second concern was how to best transition the BLOCKHASH resolution logic post fork by:

1. Either waiting for  `HISTORY_SERVE_WINDOW` blocks for the entire relevant history to persist
2. Storing of all last `HISTORY_SERVE_WINDOW` block hashes on the fork block.

> We choose to go with later as it alleviates the need to detect fork activation height to transition to the new logic in backward compatible manner as the entire requisite history will be available from the first block of the fork itself. The cost of doing so is marginal considering the `HISTORY_SERVE_WINDOW` being small.

Reviewer 3: I agree and I would honestly think (1) would create an anomaly which we do not want.

## Backwards Compatibility

> The behavior of `BLOCKHASH` opcode remains same and doesn't affect backward compatibility with the contracts deployed or the gas consumption costs as the resolution of the opcode is handled "directly" by the clients

Reviewer 2: ```suggestion
The behavior of `BLOCKHASH` opcode remains same and doesn't affect backward compatibility with the contracts deployed or the gas consumption costs as the resolution of the opcode is handled "directly" by the clients.
```

> The behavior of `BLOCKHASH` opcode remains same and this EIP doesn't affect backward compatibility with the contracts deployed or the gas consumption costs as the resolution of the opcode is handled "directly" by the clients.

Reviewer 3: Is this true? Now `BLOCKHASH` will also invoke some witness costs - right?

Author: ahh yes, the accesses need to be bundled for stateless clients (will add this) but we can keep the gas same since BLOCKHASH can ideally be resolved/served from maintaining last 256 hashes in memory

### [EIP-158](./eip-158.md) handling

This address is currently exempt from [EIP-158](./eip-158.md) cleanup in Kaustinen Verkle Testnet but we plan to address this in the following way:

* Deploy a contract à la [EIP-4788](./eip-4788.md) which just supports `get` method to resolve the BLOCKHASH as per the logic defined above (and use the generated address as the BLOCKHASH contract address).
* While the clients are expected to directly read from state (or maintain and server from memory) to resolve BLOCKHASH opcode, this contract's `get` could be invoked by transaction (via contract or otherwise) leading to a normal contract execution (and gas consumption) as per the semantics of the contract call.

## Rationale

Very similar ideas were proposed before in [EIP-210](./eip-210.md) et al. This EIP is a simplification, removing two sources of needless complexity:

1. Having a tree-like structure with multiple layers as opposed to a single list
2. Writing the EIP in EVM code
3. Serial unbounded storage of hashes for a deep access to the history

> However after weighing pros and cons, we decided to go with just a limited ring buffer to only serve the requiste `HISTORY_SERVE_WINDOW` as [EIP 4788](./eip04788) and beacon state accumulators allow (albit a bit more complex) proof against any ancestor since merge.

Reviewer: ```suggestion
However after weighing pros and cons, we decided to go with just a limited ring buffer to only serve the requiste `HISTORY_SERVE_WINDOW` as [EIP-4788](./eip04788) and beacon state accumulators allow (albit a bit more complex) proof against any ancestor since merge.
```

## Reference Implementation

 * PR 28878 of go-ethereum

>  * Active on verkle-gen-devet-4 for its verkle implementation

Reviewer 2: Two notes:
- s/devet/devnet
- I think this should say 5 now.

## Security Considerations

> Having contracts (system or otherwise) with hot update paths (branches) poses a risk of "branch" poisioning attracks where one could sprinkle trivial amounts of eth around these hot paths (branches). But it has been deemed that cost of attack would escalate significantly to cause any meaningful slow down of state root updates.

Reviewer: ```suggestion
Having contracts (system or otherwise) with hot update paths (branches) poses a risk of "branch" poisioning attacks where one could sprinkle trivial amounts of eth around these hot paths (branches). But it has been deemed that cost of attack would escalate significantly to cause any meaningful slow down of state root updates.
```
